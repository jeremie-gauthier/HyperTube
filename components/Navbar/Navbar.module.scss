@import "@/styles/theme.scss";

.container {
  @apply flex-1 h-16 p-4 bg-black items-center justify-between text-white;

  .items {
    @apply flex-1 space-x-6 text-sm ml-8 h-6 items-center;
  }

  .settings {
    @apply space-x-4;
  }

  .searchGroup {
    @apply items-center space-x-2 border border-white p-1;

    input {
      @apply text-sm bg-black outline-none;
    }
  }
}

.brandLogo {
  @apply relative cursor-pointer h-8 w-4 md:w-28;

  @media (max-width: 768px) {
    background: no-repeat center/100% url("/img/hypertube-short.png");
  }

  @media (min-width: 768px) {
    background: no-repeat center/100% url("/img/hypertube.png");
  }
}

.countryFlag {
  @apply h-8 w-8 rounded;
}

.floater {
  @apply absolute z-40 top-8 right-0 pt-4;

  &::before {
    @apply absolute arrow-up top-2 right-2;

    content: "";
  }

  .floaterContent {
    @apply relative;
    @apply bg-grey-dark opacity-75 text-sm min-w-max border border-grey-light;

    a {
      @apply font-bold space-x-2 hover:bg-black pl-3 pr-6 py-2;
    }

    button {
      @apply flex flex-row items-center space-x-2 hover:bg-black pl-3 pr-6 py-2;
    }
  }
}

.dropdownMenu {
  @apply absolute z-40 top-16 left-0 right-0 w-screen bg-grey-dark py-4 text-xl;

  a {
    @apply py-4 pl-10;

    &:hover {
      @apply bg-black;
    }
  }

  hr {
    @apply border-black;
  }

  .linkWithIcon {
    @apply flex flex-row items-center space-x-4;
  }

  .activeLink {
    @apply bg-black;
  }
}
