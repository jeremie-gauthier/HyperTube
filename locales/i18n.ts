import i18n from "i18next";
import { initReactI18next } from "react-i18next";

// This will probably help me later
// https://stackoverflow.com/questions/58277973/how-to-type-check-i18n-dictionaries-with-typescript

import componentsFR from "./translations/fr/components.json";
import componentsEN from "./translations/en/components.json";
import componentsES from "./translations/es/components.json";
import componentsJA from "./translations/ja/components.json";

import commonFR from "./translations/fr/common.json";
import commonEN from "./translations/en/common.json";
import commonES from "./translations/es/common.json";
import commonJA from "./translations/ja/common.json";

const LANGUAGES = {
  fr: {
    translation: { pages: {}, components: componentsFR, common: commonFR },
  },
  en: {
    translation: { pages: {}, components: componentsEN, common: commonEN },
  },
  es: {
    translation: { pages: {}, components: componentsES, common: commonES },
  },
  ja: {
    translation: { pages: {}, components: componentsJA, common: commonJA },
  },
};

export const langs = {
  en: "common.lang.english",
  fr: "common.lang.french",
  es: "common.lang.spanish",
  ja: "common.lang.japanese",
};

i18n.use(initReactI18next).init({
  resources: LANGUAGES,
  lng: "en",
  fallbackLng: "en",

  interpolation: {
    escapeValue: false,
  },
});

export default i18n;
