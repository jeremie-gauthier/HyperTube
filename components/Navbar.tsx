import React from "react";
import { FlexRow, FlexCol } from "@/components/Flex";
import { useTranslation } from "react-i18next";
import Link from "next/link";
import Image from "next/image";
import userMock from "@/tests/__mocks__/user"; // MOCK
import { getInitials } from "@/lib/helpers";
import useHover from "@react-hook/hover";
import { langs } from "@/locales/i18n";
import CountryFlag from "@/components/CountryFlag";
import ActiveLink from "@/components/ActiveLink";
import Magnifier from "../public/icons/magnifier.svg";
import styles from "./Navbar.module.scss";

export default function Navbar(): JSX.Element {
  return (
    <FlexRow className={styles.container}>
      <BrandLogo />

      <FlexRow className={styles.items}>
        <Navigation />
        <SearchInput />
      </FlexRow>

      <FlexRow className={styles.settings}>
        <UserMenu />
        <LangFlag />
      </FlexRow>
    </FlexRow>
  );
}

// The <a> markup is mandatory here to avoid errors
// https://github.com/vercel/next.js/issues/7915
const BrandLogo = () => (
  <Link href="/">
    <a href="/" className="flex">
      <Image
        src="/icons/hypertube.png"
        alt="Hypertube logo"
        height={31}
        width={115}
        priority
        className="cursor-pointer"
      />
    </a>
  </Link>
);

const Navigation = () => {
  const { t } = useTranslation();

  return (
    <>
      <ActiveLink href="/">{t("components.navbar.home")}</ActiveLink>
      <ActiveLink href="/movies/favorites">
        {t("components.navbar.my_list")}
      </ActiveLink>
    </>
  );
};

const SearchInput = () => {
  const { t } = useTranslation();
  // I should probably use a context here for input

  return (
    <FlexRow className={styles.searchInput}>
      <Magnifier />
      <input placeholder={t("components.navbar.search")} />
    </FlexRow>
  );
};

const UserMenu = () => {
  const { t } = useTranslation();
  const hoverableNode = React.useRef(null);
  const isHovering = useHover(hoverableNode, {
    enterDelay: 0,
    leaveDelay: 500,
  });

  return (
    <div ref={hoverableNode} className={styles.frame + " bg-red"}>
      <span>{getInitials(userMock)}</span>

      {isHovering && (
        <div className={styles.floater}>
          <div className={styles.arrow} />
          <FlexCol className={styles.floaterContent}>
            <Link href="account">{t("components.navbar.account")}</Link>
            <Link href="logout">{t("components.navbar.logout")}</Link>
          </FlexCol>
        </div>
      )}
    </div>
  );
};

const LangFlag = () => {
  const { t, i18n } = useTranslation();
  const hoverableNode = React.useRef(null);
  const isHovering = useHover(hoverableNode, {
    enterDelay: 0,
    leaveDelay: 500,
  });

  const changeLanguage = (lang: string) => {
    i18n.changeLanguage(lang);
  };

  return (
    <div ref={hoverableNode} className={styles.frame} data-testid="lang-flag">
      <CountryFlag
        lang={i18n.language}
        width={32}
        height={32}
        className="rounded"
      />

      {isHovering && (
        <div className={styles.floater}>
          <div className={styles.arrow} />
          <FlexCol className={styles.floaterContent}>
            {langs.map((lang) => (
              <button
                key={lang.value}
                type="button"
                onClick={() => changeLanguage(lang.value)}
              >
                <CountryFlag lang={lang.value} width={15} height={15} />
                <span>{t(lang.key)}</span>
              </button>
            ))}
          </FlexCol>
        </div>
      )}
    </div>
  );
};
