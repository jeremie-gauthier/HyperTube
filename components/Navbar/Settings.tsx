import React from "react";
import { FlexRow, FlexCol } from "@/components/Flex";
import { useTranslation } from "react-i18next";
import Link from "next/link";
import userMock from "@/tests/__mocks__/user"; // MOCK
import useHover from "@react-hook/hover";
import { LangSettings } from "@/components/CountryFlag";
import UserIcon from "@/components/UserIcon";
import styles from "./Navbar.module.scss";

export default function Settings(): JSX.Element {
  return (
    <FlexRow className={styles.settings}>
      <UserSettings />
      <LangSettings />
    </FlexRow>
  );
}

const UserSettings = () => {
  const hoverableNode = React.useRef(null);
  const isHovering = useHover(hoverableNode);

  return (
    <FlexCol ref={hoverableNode} className="relative">
      <UserIcon user={userMock} />

      {isHovering && (
        <div className={styles.floater}>
          <UserLinks />
        </div>
      )}
    </FlexCol>
  );
};

const UserLinks = () => {
  const { t } = useTranslation();

  return (
    <FlexCol className={styles.floaterContent}>
      <Link href="account">{t("components.navbar.account")}</Link>
      <Link href="logout">{t("components.navbar.logout")}</Link>
    </FlexCol>
  );
};
