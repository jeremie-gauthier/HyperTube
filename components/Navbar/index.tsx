/* eslint-disable jsx-a11y/no-autofocus */
import React from "react";
import { FlexCol, FlexRow } from "@/components/Flex";
import { useTranslation } from "react-i18next";
import Link from "next/link";
import Image from "next/image";
import ActiveLink from "@/components/ActiveLink";
import { RootState } from "@/state/types";
import useSelector, { useResponsiveAttribute } from "@/hooks/useSelector";
import useDispatch from "@/hooks/useDispatch";
import { setSearchInput } from "@/state/movies/actions";
import classNames from "classnames";
import userMock from "@/tests/__mocks__/user"; // MOCK
import UserIcon from "@/components/UserIcon";
import CountryFlag from "@/components/CountryFlag";
import Magnifier from "../../public/icons/magnifier.svg";
import Cross from "../../public/icons/cross.svg";
import MenuBurger from "../../public/icons/menu-burger.svg";
import styles from "./Navbar.module.scss";
import Settings from "./Settings";

export default function Navbar(): JSX.Element {
  const isTabletOrMobile = useResponsiveAttribute();

  return (
    <FlexRow className={styles.container}>
      <BrandLogo />

      {isTabletOrMobile ? <MobileView /> : <DesktopView />}
    </FlexRow>
  );
}

const MobileView = (): JSX.Element => {
  const [open, setOpen] = React.useState(false);

  return (
    <>
      <FlexRow className={`relative ${styles.items} justify-end`}>
        <SearchInput />
        {open ? (
          <Cross
            className="h-5 w-5"
            onClick={() => setOpen(false)}
            data-testid="cross-icon"
          />
        ) : (
          <MenuBurger
            className="h-5 w-5"
            onClick={() => setOpen(true)}
            data-testid="menuburger-icon"
          />
        )}
      </FlexRow>

      {open && <DropdownMenu close={() => setOpen(false)} />}
    </>
  );
};

type DropdownMenuProps = {
  close: () => void;
};

const DropdownMenu = ({ close }: DropdownMenuProps) => {
  const currentLang = useSelector((state) => state.user.lang) as string;
  const { t } = useTranslation();

  return (
    <FlexCol className={styles.dropdownMenu}>
      <NavLinks activeClassName={styles.activeLink} onClick={close} />
      <hr />
      <ActiveLink
        href="/account"
        className={styles.linkWithIcon}
        activeClassName={styles.activeLink}
      >
        <UserIcon user={userMock} />
        <span>{t("components.navbar.account")}</span>
      </ActiveLink>
      <ActiveLink
        href="/account#languages"
        className={styles.linkWithIcon}
        activeClassName={styles.activeLink}
      >
        <CountryFlag lang={currentLang} className={styles.countryFlag} />
        <span>{t("common.lang.change_language")}</span>
      </ActiveLink>
      <hr />
      <Link href="logout">{t("components.navbar.logout")}</Link>
    </FlexCol>
  );
};

const DesktopView = (): JSX.Element => (
  <>
    <FlexRow className={styles.items}>
      <NavLinks />
      <SearchInput />
    </FlexRow>

    <Settings />
  </>
);

const NavLinks = ({ ...rest }) => {
  const { t } = useTranslation();

  return (
    <>
      <ActiveLink
        activeClassName={styles.activeBorder}
        inactiveClassName={styles.inactiveBorder}
        href="/"
        {...rest}
      >
        {t("components.navbar.home")}
      </ActiveLink>
      <ActiveLink
        activeClassName={styles.activeBorder}
        inactiveClassName={styles.inactiveBorder}
        href="/movies/favorites"
        {...rest}
      >
        {t("components.navbar.my_list")}
      </ActiveLink>
    </>
  );
};

// The <a> markup is mandatory here to avoid errors
// https://github.com/vercel/next.js/issues/7915
const BrandLogo = () => {
  const isTabletOrMobile = useResponsiveAttribute();
  const logo = isTabletOrMobile
    ? "/icons/hypertube-short.png"
    : "/icons/hypertube.png";
  const width = isTabletOrMobile ? 17 : 115;

  return (
    <Link href="/">
      <a href="/" className="flex">
        <Image
          src={logo}
          alt="Hypertube logo"
          width={width}
          height={31}
          priority
          className="cursor-pointer"
        />
      </a>
    </Link>
  );
};

const SearchInput = () => {
  const { t } = useTranslation();
  const searchInput = useSelector(
    (state: RootState) => state.movie.searchInput,
  ) as string;
  const dispatch = useDispatch();
  const isVisible = searchInput.length > 0;
  const [showInput, setShowInput] = React.useState(isVisible);
  const cancelSearch = classNames({
    "cursor-pointer h-4 w-4": true,
    invisible: !isVisible,
  });

  const handleSearch = () => {
    console.log(`Request movies with name [${searchInput}]`);
  };

  const handleClose = () => {
    dispatch(setSearchInput(""));
    setShowInput(false);
  };

  return showInput ? (
    <FlexRow className={styles.searchGroup}>
      <Magnifier
        className="h-4 w-4"
        onClick={handleSearch}
        data-testid="submit-search"
      />
      <input
        placeholder={t("components.navbar.search")}
        value={searchInput}
        onChange={(evt) => dispatch(setSearchInput(evt.target.value))}
        onBlur={() => !isVisible && setShowInput(false)}
        autoFocus
      />
      <Cross
        className={cancelSearch}
        onClick={handleClose}
        data-testid="close-search"
      />
    </FlexRow>
  ) : (
    <Magnifier
      onClick={() => setShowInput(true)}
      className="h-4 w-4 cursor-pointer self-center"
      data-testid="magnifier-icon"
    />
  );
};
