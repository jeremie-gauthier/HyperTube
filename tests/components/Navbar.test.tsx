import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@/tests/__mocks__/matchMedia";
import Navbar from "@/components/Navbar";
import i18n from "@/locales/i18n";
import renderer from "react-test-renderer";

test("renders correctly", () => {
  const tree = renderer.create(<Navbar />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe("Navbar", () => {
  const mockChangeLanguage = jest.spyOn(i18n, "changeLanguage");

  test("links", () => {
    render(<Navbar />);

    const nextLink = screen.getByText(/Home/i) as HTMLLinkElement;
    expect(nextLink.href).toEqual("http://localhost/");
  });

  test("display user menu", () => {
    render(<Navbar />);
    expect(screen.queryByText(/Account/i)).not.toBeInTheDocument();
    userEvent.hover(screen.getByText(/AD/i));

    const accountLink = screen.getByText(/Account/i) as HTMLLinkElement;
    expect(accountLink).toBeInTheDocument();
    expect(accountLink.href).toEqual("http://localhost/account");

    const logoutLink = screen.getByText(/Logout/i) as HTMLLinkElement;
    expect(logoutLink).toBeInTheDocument();
    expect(logoutLink.href).toEqual("http://localhost/logout");
  });

  test("change language", () => {
    render(<Navbar />);

    expect(screen.queryByText(/English/i)).not.toBeInTheDocument();
    userEvent.hover(screen.getByTestId("lang-flag"));
    expect(screen.getByText(/English/i)).toBeInTheDocument();
    userEvent.click(screen.getByText(/French/i));
    expect(mockChangeLanguage).toHaveBeenCalledWith("fr");
  });
});
