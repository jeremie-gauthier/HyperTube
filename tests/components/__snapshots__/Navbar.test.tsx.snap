// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="flex flex-row container"
>
  <a
    className="flex"
    href="/"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "display": "inline-block",
          "margin": 0,
          "maxWidth": "100%",
          "overflow": "hidden",
          "position": "relative",
        }
      }
    >
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "display": "block",
            "maxWidth": "100%",
          }
        }
      >
        <img
          alt=""
          aria-hidden={true}
          role="presentation"
          src="data:image/svg+xml;charset=utf-8,<svg width=\\"115\\" height=\\"31\\" xmlns=\\"http://www.w3.org/2000/svg\\" version=\\"1.1\\"/>"
          style={
            Object {
              "display": "block",
              "maxWidth": "100%",
            }
          }
        />
      </div>
      <img
        alt="Hypertube logo"
        className="cursor-pointer"
        decoding="async"
        src="/_next/image?url=%2Ficons%2Fhypertube.png&w=420&q=75"
        srcSet="/_next/image?url=%2Ficons%2Fhypertube.png&w=320&q=75 1x, /_next/image?url=%2Ficons%2Fhypertube.png&w=420&q=75 2x"
        style={
          Object {
            "border": "none",
            "bottom": 0,
            "boxSizing": "border-box",
            "display": "block",
            "height": 0,
            "left": 0,
            "margin": "auto",
            "maxHeight": "100%",
            "maxWidth": "100%",
            "minHeight": "100%",
            "minWidth": "100%",
            "padding": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "visibility": "visible",
            "width": 0,
          }
        }
      />
    </div>
  </a>
  <div
    className="flex flex-row items"
  >
    <div
      className="inactive"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      components.navbar.home
      <div
        className="border"
      />
    </div>
    <div
      className="inactive"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      components.navbar.my_list
      <div
        className="border"
      />
    </div>
    <div
      className="flex flex-row searchInput"
    >
      <svg
        fill="none"
        height="12"
        viewBox="0 0 11 12"
        width="11"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.848 10.375L8.707 8.038a.494.494 0 0 0-.365-.164h-.35a5.138 5.138 0 0 0 .945-3c0-2.692-2-4.874-4.469-4.874C2 0 0 2.182 0 4.875c0 2.692 2 4.874 4.468 4.874a4.217 4.217 0 0 0 2.75-1.031V9.1c0 .15.054.293.15.398l2.142 2.337c.202.22.529.22.729 0l.607-.663a.602.602 0 0 0 .002-.797zm-6.38-2.5c-1.518 0-2.75-1.341-2.75-3 0-1.657 1.23-3 2.75-3s2.75 1.34 2.75 3c0 1.656-1.229 3-2.75 3z"
          fill="#fff"
        />
      </svg>
      <input
        onChange={[Function]}
        placeholder="components.navbar.search"
        value=""
      />
    </div>
  </div>
  <div
    className="flex flex-row settings"
  >
    <div
      className="flex flex-col frame bg-red"
    >
      <span>
        AD
      </span>
    </div>
    <div
      className="frame"
      data-testid="lang-flag"
    >
      <svg
        className="rounded"
        height={32}
        viewBox="0 0 512 512"
        width={32}
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="#41479b"
        >
          <path
            d="M188.632 0H0v39.954L188.632 163.54zM0 136.598v52.034h79.419zM0 323.369v52.033l79.419-52.033zM512 39.954V0H323.368v163.54zM512 375.402v-52.033h-79.419zM323.368 348.46V512H512v-39.954zM512 188.632v-52.034l-79.419 52.034zM0 472.046V512h188.632V348.46z"
          />
        </g>
        <g
          fill="#f5f5f5"
        >
          <path
            d="M512 472.046v-39.928L344.465 323.369h49.496L512 399.989v-24.587l-79.419-52.033H512v-26.948H296.421V512h26.947V348.46zM0 296.421v26.948h79.419L0 375.402v37.801l138.395-89.834h49.496L0 445.332v26.714L188.632 348.46V512h26.947V296.421zM215.579 0h-26.947v163.54L0 39.954v43.725l161.684 104.953h-49.496L0 115.807v20.791l79.419 52.034H0v26.947h215.579zM512 215.579v-26.947h-79.419L512 136.598V98.314l-139.136 90.318h-49.496L512 66.185V39.954L323.368 163.54V0h-26.947v215.579z"
          />
        </g>
        <g
          fill="#ff4b55"
        >
          <path
            d="M512 296.421v-80.842H296.421V0h-80.842v215.579H0v80.842h215.579V512h80.842V296.421z"
          />
          <path
            d="M138.395 323.369L0 413.203v32.129l187.891-121.963zM344.465 323.369L512 432.118v-32.129l-118.039-76.62zM161.684 188.632L0 83.679v32.128l112.188 72.825zM372.864 188.632L512 98.314V66.185L323.368 188.632z"
          />
        </g>
      </svg>
    </div>
  </div>
</div>
`;
